app/main/forms.py

```diff
@@ -74,13 +80,73 @@ class MoveUserForm(FlaskForm):
         ])
 
 
+class NotSuffixOfBlacklist:
+    "WTForms validator, asserting that supplied value is not a prefix of any of the values present in a blacklist"
+
+    # path, relative to flask app root_path, to look for suffix blacklist files. all files found here will be read,
+    # one suffix per line
+    BLACKLIST_DIR_PATH: str = "data/public_domain_suffix_blacklist"
+
+    @staticmethod
+    def _normalized_suffix(suffix: str) -> str:
+        return (
+            # all suffixes we deal with implicitly begin with a separator - make that explicit here
+            ("" if suffix.startswith(".") else ".") + suffix.strip().lower()
+            # we want to find suffixes. sorting works by prefix, so reverse all strings before we work with them
+        )[::-1]
+
+    @staticmethod
+    def _un_normalized_suffix(normalized_suffix: str) -> str:
+        # we can't fully reverse the normalization process as it throws information away, but we can restore its
+        # presentational sanity.
+        return normalized_suffix[::-1]
+
+    @classmethod
+    def _lines_from_filepath(cls, filepath: Path) -> typing.Sequence[str]:
+        with filepath.open("r", encoding="utf-8") as f:
+            return tuple(cls._normalized_suffix(line) for line in f if line)
+
+    # this value is not populated until first access because construction depends on current_app being available
+    _blacklist: typing.Optional[typing.Sequence[str]] = None
+
+    @classmethod
+    def get_blacklist(cls) -> typing.Sequence[str]:
+        # cache blacklist class-wide
+        if cls._blacklist is None:
+            # store this as a sorted tuple so we can use bisection to find specific prefixes
+            cls._blacklist = tuple(sorted(chain.from_iterable(
+                cls._lines_from_filepath(filepath)
+                for filepath in (Path(current_app.root_path) / cls.BLACKLIST_DIR_PATH).iterdir()
+                if filepath.is_file()
+            )))
+        return cls._blacklist
+
+    def __init__(self, message):
+        self.message = message
+
+    def __call__(self, form, field) -> None:
+        blacklist = self.get_blacklist()
+        normalized_data = self._normalized_suffix(field.data)
+        i = bisect_left(blacklist, normalized_data)
+        # if (the normalized) `field.data` is a prefix of any of the (normalized) entries in the blacklist, it
+        # should sort just before the first such entry. hence we just have to check whether normalized_data is
+        # a prefix of the entry that sorts right after it. an "equal" match should work in the same way because
+        # we use bisect_left
+        if i < len(blacklist) and blacklist[i].startswith(normalized_data):
```

Comment:
I'm don't think bisect actually is quicker than an `in` operation on a set. You've already got O(n) because you're iterating the blacklist to normalise the prefixes. Then a bisect left performs an insertion, which is also pretty much O(n).

You can achieve the same complexity with 

`cls._normalized_suffix(field) in {cls._normalized_suffix(line) for line in open(filepath)}`

Or something like that

Reply:
https://stackoverflow.com/a/53023435/3967444

Reply:
Can't use `set`s because we're not doing exact matching. We want to match suffixes _of suffixes_ too (consider the blacklist containing `.foo.bar`, we want to disallow `.bar` too. Bisection is also O(log n) not O(n).

Reply:
We're also not inserting anything, only using `bisect_left` to look up the hypothetical insertion _point_.

Reply:
This confusion probably could have been avoided by not using `i` as a var name.